utils/frame/benchmarking-cli/src/bedrock/storage/write.rs

```diff
@@ -0,0 +1,175 @@
+		>,
+	{
+		let COLUMN = sc_client_db::columns::STATE;
+		let state_version = StateVersion::V0;
+
+		let block = BlockId::Number(prov.usage_info().chain.best_number);
+		info!("Best block is {}", block);
+		let header = prov.header(block).unwrap().unwrap();
+		let original_root = header.state_root();
+
+		let mut rng = StdRng::seed_from_u64(self.seed);
+
+		let state = DbState::<Block>::new(storage.clone(), *original_root);
+		info!("Preparing {} keys", state.keys(Default::default()).len());
+
+		// Load all KV pairs from the DB.
+		let mut original_kvs = state.pairs();
+		// Randomly shuffle the KV pairs.
+		original_kvs.shuffle(&mut rng);
+		let mut kvs = original_kvs.clone();
+		// Replace each value with a random value of the same size.
+		// NOTE: We use a possibly higher entropy than the original value,
+		// could be improved but acts as an overestimation which is fine for now.
+		for (_, v) in kvs.iter_mut() {
+			*v = random_vec(&mut rng, v.len());
+		}
+
+		info!("Starting run");
+		// Store the time that it took to write a value.
+		let mut time_by_key_index = Vec::<Duration>::with_capacity(kvs.len());
+		// Write each value in one commit.
+		for ((_, ov), (k, v)) in original_kvs.iter().zip(kvs.iter()) {
+			// Interesting part here:
+			let start = Instant::now();
+			// Create a TX that will modify the Trie in the DB and
+			// calculate the root hash of the Trie after appending `delta`.
+			let delta = vec![(k.clone(), v.clone())];
+			let delta = delta.iter().map(|(k2, v2)| (k2.as_ref(), Some(v2.as_ref())));
+			let (root, tx) = state.storage_root(delta, state_version);
+			let tx = convert_tx::<Block>(COLUMN, tx);
+			db.commit(tx).expect("Must commit");
+
+			time_by_key_index.push(start.elapsed());
+
+			let new_state = DbState::<Block>::new(storage.clone(), root);
+
+			// Replace it with the original value.
+			let delta = vec![(k.clone(), ov.clone())];
+			let delta = delta.iter().map(|(k2, v2)| (&**k2, Some(v2.as_ref())));
+			let (root, tx) = new_state.storage_root(delta, state_version);
+			let tx = convert_tx::<Block>(COLUMN, tx);
+			db.commit(tx).expect("Must commit");
+
+			// Replacing the value should bring us back to the original root.
+			if root != *original_root {
+				// FAILS HERE
+				log::error!("Wrong final root. {:?} vs {:?}", root, original_root);
+				std::process::exit(1);
+			}
+		}
+
+		let mut res = TimeResult::default();
+		for (i, (_, v)) in original_kvs.clone().into_iter().enumerate() {
+			res.time_by_size.push((v.len() as u64, time_by_key_index[i].as_nanos()));
+		}
+
+		let template = res.post_process(&cfg, &self.post_proc, false)?;
+		template.write(&self.post_proc.weight_out, false)
+	}
+}
+
+fn convert_tx<B: BlockT>(
+	col: ColumnId,
+	mut tx: sp_trie::PrefixedMemoryDB<HashFor<B>>,
+) -> Transaction<H256> {
+	let mut ret = sp_database::Transaction::<H256>::default();
+
+	for (mut k, (v, rc)) in tx.drain().into_iter() {
+		// HACK
+		k.drain(0..k.len() - 32 /* sc_client_db::DB_HASH_LEN */);
```

Comment:
def put a comment here :)

cc @cheme

Reply:
@ggwpez, reminder that this works for paritydb, but not for rocksdb.

Reply:
I put a comment. Looks still a bit ominous.  
Maybe we could add a `sanitize_key` or `canonicalize_key` function somewhere in the DB crate.